pkg/controller/garbagecollector/graph_builder.go

```diff
@@ -0,0 +1,362 @@
+// "BlockOwnerDeletion=false", add the object to the dirty queue.
+func (p *GraphBuilder) addUnblockedOwnersToDirtyQueue(removed []metav1.OwnerReference, changed []ownerRefPair) {
+	for _, ref := range removed {
+		if ref.BlockOwnerDeletion != nil && *ref.BlockOwnerDeletion {
+			node, found := p.uidToNode.Read(ref.UID)
+			if !found {
+				glog.V(5).Infof("cannot find %s in uidToNode", ref.UID)
+				continue
+			}
+			p.attemptToDelete.Add(node)
+		}
+	}
+	for _, c := range changed {
+		if c.old.BlockOwnerDeletion != nil && *c.old.BlockOwnerDeletion &&
+			c.new.BlockOwnerDeletion != nil && !*c.new.BlockOwnerDeletion {
+			node, found := p.uidToNode.Read(c.new.UID)
+			if !found {
+				glog.V(5).Infof("cannot find %s in uidToNode", c.new.UID)
+				continue
+			}
+			p.attemptToDelete.Add(node)
+		}
+	}
+}
+
+// Dequeueing an event from graphChanges, updating graph, populating dirty_queue.
+func (p *GraphBuilder) processEvent() {
+	item, quit := p.graphChanges.Get()
+	if quit {
+		return
+	}
+	defer p.graphChanges.Done(item)
+	event, ok := item.(*event)
+	if !ok {
+		utilruntime.HandleError(fmt.Errorf("expect a *event, got %v", item))
+		return
+	}
+	obj := event.obj
+	accessor, err := meta.Accessor(obj)
+	if err != nil {
+		utilruntime.HandleError(fmt.Errorf("cannot access obj: %v", err))
+		return
+	}
+	typeAccessor, err := meta.TypeAccessor(obj)
+	if err != nil {
+		utilruntime.HandleError(fmt.Errorf("cannot access obj: %v", err))
+		return
+	}
+	glog.V(5).Infof("GraphBuilder process object: %s/%s, namespace %s, name %s, event type %s", typeAccessor.GetAPIVersion(), typeAccessor.GetKind(), accessor.GetNamespace(), accessor.GetName(), event.eventType)
+	// Check if the node already exsits
+	existingNode, found := p.uidToNode.Read(accessor.GetUID())
+	switch {
+	case (event.eventType == addEvent || event.eventType == updateEvent) && !found:
+		newNode := &node{
+			identity: objectReference{
+				OwnerReference: metav1.OwnerReference{
+					APIVersion: typeAccessor.GetAPIVersion(),
+					Kind:       typeAccessor.GetKind(),
+					UID:        accessor.GetUID(),
+					Name:       accessor.GetName(),
+				},
+				Namespace: accessor.GetNamespace(),
+			},
+			dependents:         make(map[*node]struct{}),
+			owners:             accessor.GetOwnerReferences(),
+			deletingDependents: beingDeleted(accessor) && hasDeleteDependentsFinalizer(accessor),
+			beingDeleted:       beingDeleted(accessor),
+		}
+		p.insertNode(newNode)
+		// the underlying delta_fifo may combine a creation and deletion into one event
+		if shouldOrphanDependents(event, accessor) {
+			glog.V(5).Infof("add %s to the attemptToOrphan", newNode.identity)
+			p.attemptToOrphan.Add(newNode)
+		}
+		if startsWaitingForDependentsDeletion(event, accessor) {
+			glog.V(2).Infof("add %s to the attemptToDelete, because it's waiting for its dependents to be deleted", newNode.identity)
+			p.attemptToDelete.Add(newNode)
```

Comment:
if we're going to wait, why not do this after you queue the others (reorder with the loop). Not that its likely to matter, but this just feel backwards to my expectations.